lisa/runner.py

```diff
@@ -257,77 +287,69 @@ def _start_loop(self) -> None:
-                                f"fetched task from {runner.id}: '{task_message}'"
-                            )
-                            task_manager.submit_task(task)
-                        else:
-                            # current runner may not be done, but it doesn't
-                            # have task temporarily. The root runner can start
-                            # tasks from next runner.
-                            break
-                        if not task_manager.has_idle_worker():
-                            has_idle_worker = False
-                            break
-                    if runner.is_done:
-                        # remove fully completed runner.
-                        runner.close()
-                        remaining_runners.remove(runner)
+                if task_manager.has_idle_worker():
+                    if has_more_runner:
+                        # add new runner upto max concurrency if idle workers
+                        # are available
+                        try:
+                            while True:
+                                remaining_runners.append(next(runner_iterator))
+                                if len(remaining_runners) > self._max_concurrency:
+                                    break
+                                self._log.debug(
+                                    f"Added runner {remaining_runners[-1].id}"
+                                )
+                        except StopIteration:
+                            has_more_runner = False
+                    else:
+                        # reduce CPU utilization from infinite loop when idle
+                        # workers are present but no task to run.
                         self._log.debug(
-                            f"runner '{runner.id}' is done, "
-                            f"remaining runners {[x.id for x in remaining_runners]}"
+                            "Idle worker available but no runner available..."
                         )
-                        break
-                    if not has_idle_worker:
-                        # uses the result from previous runner, because the
-                        # worker status may be changed after first runner
-                        # finished. Evne in this case, it should try result from
-                        # previous runner firstly in next run.
-                        break
-
-                while (
-                    len(remaining_runners) < self._max_concurrency and has_more_runner
-                ):
-                    # Fetch runners, if runner count is smaller than concurrency
-                    # count. It makes sure all concurrency can run.
-                    try:
-                        remaining_runners.append(next(runner_iterator))
-                    except StopIteration:
-                        has_more_runner = False
+                        time.sleep(1)
+
+                if not remaining_runners:
```

Comment:
Which condition is this needed? If so, add comments here.

Reply:
This is to avoid infinite loop when idle workers are available but not items in in remaining runners

Reply:
If there is no remaining runner, that means all runners are done. There shouldn't have any running task even. If there are running task, but remaining runner is empty, it should be an assertion error. So this condition may not effective.

Reply:
We may add a check that, `not task_manager.has_idle_worker() and not has_more_runner`